### PR TITLE
A new word in visibility detection

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
This PR speeds up visibility detection in general, fixes it for landscape mode and also fixes scrolling to element for Xcode9+
Previously visibility detection was based on _hittable_ property value, but this property does not work for static elements, like an image in a cell, since hittablity for such elements is defined by gesture recogniser. Now we use another approach: we detect the visibility of the particular element based on the intersection between its own frame and all ancestor frames. This gives better results, since even if the element is not hittable the viability value is calculated properly for it:
before
now

As a result, the issues like https://github.com/facebook/WebDriverAgent/issues/746 should be addresses after the PR is merged and the overall visibility detection speed should be better.

### CODE REVIEW

**Comment by mykola-mokhnach** [comment]
@antiarchit Scrolling has been fixes, because I've actually noticed this bug while running visibility integration tests on my machine with Xcode9:
> And regarding the scrolling issue: Xcode9 always puts nil into self.application property of all XCElementSnapshot instances. That is why we need to explicitly pass the current application instance from the corresponding XCUIElement instance, where it is defined.

But yep, it would be better to extract this fix into a separate PR

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m [hunk 1]
```diff
@@ -30,35 +27,36 @@ - (BOOL)fb_isVisible
 
 @implementation XCElementSnapshot (FBIsVisible)
 
++ (BOOL)fb_isVisibleInContainerWindow:(XCElementSnapshot *)window selfElement:(XCElementSnapshot *)element hierarchyIntersection:(nullable NSValue *)irect appFrame:(CGRect)appFrame
+{
+  XCElementSnapshot *parent = element.parent;
+  CGRect intersectionRect = nil == irect ?
```
Comment (antiarchit) [inline WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: `irect ?` should do it?

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m [hunk 2]
```diff
@@ -30,35 +27,36 @@ - (BOOL)fb_isVisible
 
 @implementation XCElementSnapshot (FBIsVisible)
 
++ (BOOL)fb_isVisibleInContainerWindow:(XCElementSnapshot *)window selfElement:(XCElementSnapshot *)element hierarchyIntersection:(nullable NSValue *)irect appFrame:(CGRect)appFrame
```
Comment (antiarchit) [inline WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: `intersectionRectangle` instead?

**Comment by mykola-mokhnach** [comment]
@antiarchit nope, scrolling changes were moved to https://github.com/facebook/WebDriverAgent/pull/806

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m (hunk 2 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: renamed

File: WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m (hunk 1 above)
Comment (mykola-mokhnach) [reply WebDriverAgentLib/Categories/XCUIElement+FBIsVisible.m]: renamed

**Comment by antiarchit** [comment]
Still showing up in the [commit](https://github.com/facebook/WebDriverAgent/pull/794/commits/dfe0bb60ab5376546f8ba0910de4fb25f58f1ee3).

Also, I wonder if `fb_isVisibleInContainerWindow:selfElement:hierarchyIntersection:appFrame:` would be better as an instance method on `XCElementSnapshot`. The parameter `selfElement` seems extraneous.

**Comment by mykola-mokhnach** [comment]
No problems, I'll squash it. Regarding the instance method - it is not the case here, as the `element` value is different on different recursion level and contains ancestor elements of the original one

**Comment by antiarchit** [comment]
> Regarding the instance method - it is not the case here, as the element value is different on different recursion level and contains ancestor elements of the original one

Sure, but you can call it on the parent after the parent is computed?

**Comment by mykola-mokhnach** [comment]
> Sure, but you can call it on the parent after the parent is computed?

Perhaps then I'll have to replace recursion with a loop. Would you prefer this approach?

**Comment by antiarchit** [comment]
> Perhaps then I'll have to replace recursion with a loop. Would you prefer such approach?

The same recursion would work. You'll be calling the method on `parent` instead of `self.class`, that's all.

**Comment by mykola-mokhnach** [comment]
Updated. Thanks for the hint

**Pull request closed without merging**